source/octf/trace/parser/IoTraceEventHandler.cpp

```diff
@@ -0,0 +1,142 @@
+        auto device = traceEvent->io().deviceid();
+        // Allocate new parsed IO event in the cache
+        auto &cachedEvent = m_cache[sid];
+        cachedEvent.mutable_header()->CopyFrom(traceEvent->header());
+
+        {
+            auto &dst = *cachedEvent.mutable_io();
+            const auto &src = traceEvent->io();
+
+            dst.set_lba(src.lba());
+            dst.set_len(src.len());
+            dst.set_ioclass(src.ioclass());
+            dst.set_deviceid(src.deviceid());
+            dst.set_operation(src.operation());
+            dst.set_flush(src.flush());
+            dst.set_fua(src.fua());
+
+            dst.set_qd(++m_ioQueueDepth);
+        }
+
+        cachedEvent.mutable_device()->set_name(m_devices[device].name());
+    } break;
+
+    case Event::EventTypeCase::kIoCompletion: {
+        if (m_ioQueueDepth) {
+            m_ioQueueDepth--;
+        }
+
+        auto lba = traceEvent->mutable_iocompletion()->lba();
+        auto len = traceEvent->mutable_iocompletion()->len();
+        auto timestamp = traceEvent->mutable_header()->timestamp();
+        auto sid = traceEvent->mutable_header()->sid();
+        auto error = traceEvent->mutable_iocompletion()->error();
+        // Find in cache which IO has been completed. We match by LBA and
+        // length
+        auto iter = m_cache.begin();
+        auto end = m_cache.end();
+        for (; iter != end; ++iter) {
+            if (iter->first > sid) {
+                // Iterating over higher SID, what means over never event,
+                // so
+                break;
+            }
+
+            if (iter->second.mutable_io()->lba() == lba &&
+                iter->second.mutable_io()->len() == len) {
+                uint64_t latency =
+                        timestamp - iter->second.mutable_header()->timestamp();
+
+                // IO found, set latency and result of IO
+                iter->second.mutable_io()->set_latency(latency);
+                iter->second.mutable_io()->set_error(error);
+
+                break;
+            }
+        }
+        flushEvetns();
+    } break;
+
+    case Event::kFilesystemMeta: {
+        auto sid = traceEvent->mutable_filesystemmeta()->refsid();
+
+        auto iter = m_cache.find(sid);
+        if (iter != m_cache.end()) {
+            auto &dst = *iter->second.mutable_file();
+            const auto &src = *traceEvent->mutable_filesystemmeta();
+            dst.set_id(src.fileid());
+            dst.set_offset(src.fileoffset());
+            dst.set_size(src.filesize());
+        }
+    } break;
+
+    default: { } break; }
+}
+
+std::shared_ptr<proto::trace::Event>
+IoTraceEventHandler::getEventMessagePrototype() {
+    return std::make_shared<proto::trace::Event>();
+}
+
+void IoTraceEventHandler::flushEvetns() {
+    auto iter = m_cache.begin();
+    auto end = m_cache.end();
+    for (; iter != end;) {
+        if (iter->second.mutable_io()->latency()) {
```

Comment:
Please use normal getter.
Please also provide comment of what is this 'if' checking (checking for completion presence?)